Cleanup guestinfo keys such as userdata

akutz opened the pull request:
This patch adds the ability to clear data from the `guestinfo` dictionary after the data has been read by the datasource. This feature is enabled by adding the following to the metadata:

```yaml
cleanup-guestinfo:
- userdata
- vendordata
```
    
When the above snippet is added to a VM's metadata, the following keys will all be set to a single space character:

*  `guestinfo.userdata`
* `guestinfo.userdata.encoding`
* `guestinfo.vendordata`
* `guestinfo.vendordata.encoding`

cc @jessehu

akutz committed:
Add ability to clear guestinfo keys after reading

This patch adds the ability to clear data from the guestinfo dictionary
after the data has been read by the datasource. This feature is enabled
by adding the following to the metadata:

        cleanup-guestinfo:
        - userdata
        - vendordata

When the above snippet is added to a VM's metadata, the keys
"guestinfo.userdata", "guestinfo.userdata.encoding",
"guestinfo.vendordata", and "guestinfo.vendordata.encoding" will all be
set to the value of a single space character.

akutz committed:
Update README

This patch updates the README to bring it up to speed with the most
recent datasource features.

jessehu commented on DataSourceVMwareGuestInfo.py:
```diff
@@ -106,7 +106,8 @@ def get_data(self):
-        if not get_data_access_method():
+        data_access_method = get_data_access_method()
+        if not data_access_method:
             LOG.error("vmware-rpctool is required to fetch guestinfo value")
```
Thanks @akutz for this patch. This comment might not related to this patch. Since data_access_method can be VMX_GUESTINFO, /path/to/vmware-rpctool, or None,  then "vmware-rpctool" in the log msg seems not accurate.

akutz replied:
Yeah, this isn't directly related. And since `VMX_GUESTINFO` is more for testing, I don't feel it's necessary to adjust the happy-path log messages to account for the env var configuration.

jessehu commented on DataSourceVMwareGuestInfo.py:
```diff
@@ -268,6 +273,61 @@ def get_guestinfo_value(key):
+    # key to a single space is as close as it gets to clearing an existing
+    # guestinfo key.
+    if value == "":
+        value = " "
```
A note for this change. Setting it to a single space doesn't break cloud-init when cloud-init runs afterwards, because cloud-init-vmware-guestinfo strips the spaces :

- https://github.com/vmware/cloud-init-vmware-guestinfo/blob/b832d4a0cc8ec359249a233dbb61f888116b7de0/DataSourceVMwareGuestInfo.py#L257
- https://github.com/vmware/cloud-init-vmware-guestinfo/blob/b832d4a0cc8ec359249a233dbb61f888116b7de0/DataSourceVMwareGuestInfo.py#L277

akutz replied:
Hi @jessehu,

I'm curious by what you mean by:

> Setting it to a single space doesn't break cloud-init 

Did you mean it **does** break the datasource? Or if you meant what you said, is there a reason you think it's worth calling out the intended behavior?

FWIW, [the data isn't cleaned up until **after** it's read](https://github.com/akutz/cloud-init-vmware-guestinfo/blob/ee087279dc3231b8c1e4372d392aade0769594a3/DataSourceVMwareGuestInfo.py#L114-L125):

```python
        # Get the metadata.
        self.metadata = load_metadata()

        # Get the user data.
        self.userdata_raw = guestinfo('userdata')

        # Get the vendor data.
        self.vendordata_raw = guestinfo('vendordata')

        # Check to see if any of the guestinfo data should be removed.
        if data_access_method == VMWARE_RPCTOOL:
            clear_guestinfo_keys(self.metadata['cleanup-guestinfo'])
```

So while I'm not sure what your intent was with your original comment, I think the PR's design is okay based on the data workflow model. Do you agree, or am I missing something super obvious?

jessehu replied:
Hi Andrew, I meant what I said :) It's a note for reviewers to understand setting the value to a single space works like a charm because the code already handles this case.

jessehu commented on DataSourceVMwareGuestInfo.py:
```diff
@@ -119,6 +120,10 @@ def get_data(self):
 
+        # Check to see if any of the guestinfo data should be removed.
+        if data_access_method == VMWARE_RPCTOOL:
+            clear_guestinfo_keys(self.metadata['cleanup-guestinfo'])
```
Do we need to run `vmware-rpctool 'info-set guestinfo.userdata.cleared true'` after clear_guestinfo_keys() to explicitly indicate the userdata is readed and cleared? We need to tell CAPV this userdata is readed and should not be set again. Although CAPV can check if guestinfo.userdata equals to a single space to find out if userdata is cleared by cloud-init, but a dedicated key like 'guestinfo.userdata.cleared' seems more clear. Maybe vmware-rpctool will support removing a key in future version, then we don't need to set guestinfo.userdata to a single space instead of removing it.

akutz replied:
Hi @jessehu,

I don't think it's necessary to have an additional field to indicate the data has been cleared by the data-source. If CAPV or any other client really wants to know, all that's necessary is to inspect the data to see if it's empty or a single white-space character.

As for ensuring it's not set more than once, [CAPV assigns the cloud-init userdata during a clone operation](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/025bdc51fcc9d7260522b0927f2c3a73675109c0/pkg/services/govmomi/vcenter/clone.go#L45-L46), which occurs only once per machine.

For other clients I *could* see the argument that it's nice to know if the data was explicitly cleared. Instead of a separate key, however, I'm inclined to set the value to something like `---` or something to indicate the data has been read. That way:

* the value is till valid YAML
* it's evident the data has been read and cleared

Thoughts?

akutz replied:
Hi @jessehu,

I updated the clear value to `---`. A key's corresponding encoding key is set to the single whitespace character. This ensures the next time a cleared key is read, all that will be loaded is an empty YAML document.

jessehu replied:
Thanks for the change. '---' is a clear indication for emtpy value 👍

akutz committed:
Updated clear value for guestinfo keys

This patch updates the behavior of clearing a key from the guestinfo
key-space. For example, if the "userdata" key is cleared then
"guestinfo.userdata" is set to "---" and "guestinfo.userdata.encoding"
is set to an empty string.

akutz commented:
This PR was tested with a CentOS image at https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/582#issuecomment-563020529, and it works. I plan to merge the PR after I test it with an Ubuntu-based image.

akutz commented:
It works with Ubuntu as well. Merging the PR.
